python/cudf/cudf/core/frame.py

```diff
@@ -1160,17 +1163,34 @@ def _positions_from_column_names(self, column_names):
             if name in set(column_names)
         ]
 
-    def _copy_type_metadata(self: T, other: T) -> T:
+    def _copy_type_metadata(
+        self: T,
+        other: T,
+        *,
+        override_dtypes: Optional[abc.Iterable[Optional[Dtype]]] = None,
+    ) -> T:
         """
         Copy type metadata from each column of `other` to the corresponding
         column of `self`.
+
+        If override_dtypes is provided, any non-None entry
+        will be used in preference to the relevant column of other to
+        provide the new dtype.
+
         See `ColumnBase._with_type_metadata` for more information.
         """
-        for name, col, other_col in zip(
-            self._data.keys(), self._data.values(), other._data.values()
+        for (name, col), dtype in zip(
+            self._data.items(),
+            (
```

Comment:
This generator is pretty nontrivial. We should give it a name and use that name in the enclosing `zip` call. This section feels a bit "too clever", with multiple levels of `zip`, and `or` used to handle potential `None` values twice in the same generator. Also I tend to avoid `or` as way to handle `None` values, because it is typically unclear if other falsy types are possible without knowing (or assuming the type of) the input data. Maybe do this:

```python
# Handle this case up front
if override_dtypes is None:
    override_dtypes = itertools.repeat(None)
# Generate the dtypes
dtypes = [
    dtype if dtype is not None else col.dtype
    for (dtype, col) in zip(override_dtypes, other._data.values())
]
# Use this in the bigger zip:
for (name, col), dtype in zip(self._data.items(), dtypes):
    ...
```

Reply:
> Also I tend to avoid or as way to handle None values,

I hate the lack of proper option types in python :(. FWIW, in this case, falsy values would have been fine since dtypes can't be false and `zip(() or repeat(None), ()) == zip((), ())`